README.md
# bufr2json

[![Build Status](https://badges.herokuapp.com/travis/ARPA-SIMC/bufr2json?branch=master&env=DOCKER_IMAGE=centos:7&label=centos7)](https://travis-ci.org/ARPA-SIMC/bufr2json)
[![Build Status](https://badges.herokuapp.com/travis/ARPA-SIMC/bufr2json?branch=master&env=DOCKER_IMAGE=fedora:27&label=fedora27)](https://travis-ci.org/ARPA-SIMC/bufr2json)
[![Build Status](https://badges.herokuapp.com/travis/ARPA-SIMC/bufr2json?branch=master&env=DOCKER_IMAGE=fedora:28&label=fedora28)](https://travis-ci.org/ARPA-SIMC/bufr2json)
[![Build Status](https://badges.herokuapp.com/travis/ARPA-SIMC/bufr2json?branch=master&env=DOCKER_IMAGE=fedora:29&label=fedora29)](https://travis-ci.org/ARPA-SIMC/bufr2json)
[![Build Status](https://badges.herokuapp.com/travis/ARPA-SIMC/bufr2json?branch=master&env=DOCKER_IMAGE=fedora:rawhide&label=fedorarawhide)](https://travis-ci.org/ARPA-SIMC/bufr2json)
[![Build Status](https://copr.fedorainfracloud.org/coprs/simc/stable/package/bufr2json/status_image/last_build.png)](https://copr.fedorainfracloud.org/coprs/simc/stable/package/bufr2json/)

Convert BUFR data to GeoJSON


NEWS
Version 0.14 - 2015-09-08

- Dballe output format
- Removed options

Version 0.11 - 2015-06-17

- dballe-7.1-4828 compatible

Version 0.10 - 2015-05-25

- dballe-7.1-4737 compatible

Version 0.9 - 2015-05-25

- C++11 compatibility
- Removed default boolean options
- Changed --attributes option (print all the attributes)

Version 0.8 - 2014-02-25

- "ident" property

Version 0.7.1 - 2014-02-19

- Fixed segfault

Version 0.7 - 2014-02-19

- Optional geohash size

Version 0.6 - 2014-02-06

- "lon" and "lat" properties
- Fixed autoconf library check

Version 0.5 - 2013-07-09

- Invalid data (B33196!=0) are skipped by default

- Option `--skip-invalid` and `--no-skip-invalid`

Version 0.4 - 2013-07-05

- libyajl2 support

- Option `--attributes` print the following data attributes:
  * B33192: climate_consistency
  * B33193: time_consistency
  * B33194: space_consistency
  * B33196: invalidated
  * B33197: manual_replacement

- Option `--geohash` print the geohash (length:12) (source:
  https://github.com/lyokato/libgeohash)


bufr2json/dump.h
/*
 * Copyright (C) 2015  ARPA-SIM <urpsim@smr.arpa.emr.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Author: Emanuele Di Giacomo <edigiacomo@arpa.emr.it>
 */
#ifndef BUFR2JSON_DUMP_H
#define BUFR2JSON_DUMP_H

#include <ostream>

#include <dballe/cmdline/processor.h>

#include <bufr2json/json.h>


namespace bufr2json {
namespace dump {

class JsonDumper : public dballe::cmdline::Action {
 protected:
    std::ostream& out;
    json::YajlGenerator json;

 public:
    JsonDumper(std::ostream& out);
    virtual void start_dump() = 0;
    virtual void end_dump() = 0;
    void flush();
};

class GeoJsonDumper : public JsonDumper {
 public:
    using JsonDumper::JsonDumper;
    virtual void start_dump() override;
    virtual void end_dump() override;
    virtual bool operator()(const dballe::cmdline::Item& item) override;
};

class DballeJsonDumper : public JsonDumper {
 public:
    using JsonDumper::JsonDumper;
    virtual void start_dump() override;
    virtual void end_dump() override;
    virtual bool operator()(const dballe::cmdline::Item& item) override;
};

}
}

#endif


bufr2json/json.h
/*
 * Copyright (C) 2015  ARPA-SIM <urpsim@smr.arpa.emr.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Author: Emanuele Di Giacomo <edigiacomo@arpa.emr.it>
 */
#ifndef BUFR2JSON_JSON_H
#define BUFR2JSON_JSON_H

#include <string>

#include <yajl/yajl_gen.h>

#include <wreport/var.h>
#include <dballe/types.h>


namespace bufr2json {
namespace json {

/**
 * Simple C++ wrapper for yajl_gen
 */
class YajlGenerator {
 private:
    yajl_gen json;

 public:
    YajlGenerator(bool beautify, const char* indent=NULL);
    ~YajlGenerator();
    std::string flush();
    void start_list();
    void end_list();
    void start_map();
    void end_map();
    void add_null();
    void add_bool(bool val);
    void add_int(long int val);
    void add_double(double val);
    void add_string(const std::string& val);
    void add_number(const std::string& val);
    void add_int_or_null(long int val);
    void add_var_value(const wreport::Var& var);
};

}
}

#endif


src/bufr2json.cc
/*
 * Copyright (C) 2013-2015  ARPA-SIM <urpsim@smr.arpa.emr.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Author: Emanuele Di Giacomo <edigiacomo@arpa.emr.it>
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <cstdlib>
#include <iostream>
#include <sstream>

#include <dballe/core/var.h>
#include <dballe/msg/msg.h>
#include <dballe/msg/context.h>
#include <dballe/cmdline/processor.h>

#include <yajl/yajl_version.h>
#include <yajl/yajl_gen.h>

#include <bufr2json/json.h>
#include <bufr2json/dump.h>


#include <getopt.h>

void show_help(std::ostream& out) {
    out << "Usage: bufr2json [OPTION]... [FILE]..."  << std::endl
        << "Convert BUFR files to JSON format" << std::endl
        << std::endl
        << " -f,--format=FORMAT     JSON output format (geojson, dballe. Default: geojson)" << std::endl
        << " -h,--help           show this help and exit" << std::endl
        << " -V,--version        show version and exit" << std::endl
        << std::endl
        << "With no FILE read standard input" << std::endl;
}

void show_version(std::ostream& out) {
    out << "bufr2json " << PACKAGE_VERSION << std::endl;
}

int main(int argc, char **argv)
{
    struct {
        std::string format;
    } input_options = {
        "geojson",
    };
    std::list<std::string> inputlist;

    while (1) {
        static struct option long_options[] = {
            {"format",          required_argument,0,                     'f'  },
            {"help",            no_argument,0,                           'h'  },
            {"version",         no_argument,0,                           'V'  },
            {0, 0, 0, 0 }
        };
        int option_index = 0;
        int c = getopt_long(argc, argv, "hVf:", long_options, &option_index);
        if (c == -1) break;
        switch (c) {
            case 0: break;
            case 'h': show_help(std::cout); return 0;
            case 'V': show_version(std::cout); return 0;
            case 'f': {
                input_options.format = optarg;
                break;
            }
            default: show_help(std::cerr); return 1;
        }
    }
    while (optind < argc) {
        std::string f(argv[optind++]);
        if (f == "-") f = "";
        inputlist.push_back(f);
    }

    dballe::cmdline::Reader reader;

    using bufr2json::dump::JsonDumper;
    using bufr2json::dump::GeoJsonDumper;
    using bufr2json::dump::DballeJsonDumper;

    std::unique_ptr<JsonDumper> dumper;
    if (input_options.format == "geojson")
        dumper.reset(new GeoJsonDumper(std::cout));
    else if (input_options.format == "dballe")
        dumper.reset(new DballeJsonDumper(std::cout));
    else {
        std::cerr << "Invalid JSON format: " << input_options.format << std::endl;
        return 1;
    }

    dumper->start_dump();
    reader.read(inputlist, *dumper);
    dumper->end_dump();
    dumper->flush();

    return 0;
}


bufr2json/json.cc
/*
 * Copyright (C) 2015  ARPA-SIM <urpsim@smr.arpa.emr.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Author: Emanuele Di Giacomo <edigiacomo@arpa.emr.it>
 */
#include <bufr2json/json.h>

#include <wreport/varinfo.h>

#include <yajl/yajl_version.h>

namespace bufr2json {
namespace json {

YajlGenerator::YajlGenerator(bool beautify, const char* indent) {
#if YAJL_MAJOR == 1
    yajl_gen_config conf = {beautify, indent};
    json = yajl_gen_alloc(&conf, NULL);
#elif YAJL_MAJOR == 2
    json = yajl_gen_alloc(NULL);
    yajl_gen_config(json, yajl_gen_beautify, beautify);
    if (indent)
        yajl_gen_config(json, yajl_gen_indent_string, indent);
#endif
}
YajlGenerator::~YajlGenerator() {
    yajl_gen_free(json);
}
std::string YajlGenerator::flush() {
    const unsigned char* buf;
#if YAJL_MAJOR == 1
    unsigned int len;
#elif YAJL_MAJOR == 2
    size_t len;
#endif
    yajl_gen_get_buf(json, &buf, &len);
    std::string str((const char*)buf);
    yajl_gen_clear(json);
    return str;
}
void YajlGenerator::start_list() {
    yajl_gen_array_open(json);
}
void YajlGenerator::end_list() {
    yajl_gen_array_close(json);
}
void YajlGenerator::start_map() {
    yajl_gen_map_open(json);
}
void YajlGenerator::end_map() {
    yajl_gen_map_close(json);
}
void YajlGenerator::add_null() {
    yajl_gen_null(json);
}
void YajlGenerator::add_bool(bool val) {
    yajl_gen_bool(json, val);
}
void YajlGenerator::add_int(long int val) {
    yajl_gen_integer(json, val);
}
void YajlGenerator::add_double(double val) {
    yajl_gen_double(json, val);
}
void YajlGenerator::add_string(const std::string& val) {
    yajl_gen_string(json, (const unsigned char*)val.c_str(), val.size());
}
void YajlGenerator::add_number(const std::string& val) {
    yajl_gen_number(json, val.c_str(), val.size());
}
void YajlGenerator::add_int_or_null(long int val) {
    if (val == dballe::MISSING_INT)
        add_null();
    else
        add_int(val);
}
void YajlGenerator::add_var_value(const wreport::Var& var) {
    if (var.isset()) {
        if (var.info()->type == wreport::Vartype::String ||
            var.info()->type == wreport::Vartype::Binary) {
            add_string(var.format());
        } else {
            add_number(var.format());
        }
    } else {
        add_null();
    }
}

}
}


bufr2json/dump.cc
/*
 * Copyright (C) 2015  ARPA-SIM <urpsim@smr.arpa.emr.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Author: Emanuele Di Giacomo <edigiacomo@arpa.emr.it>
 */
#include <bufr2json/dump.h>

#include <sstream>

#include <wreport/var.h>

#include <dballe/message.h>
#include <dballe/msg/msg.h>
#include <dballe/msg/context.h>

namespace bufr2json {
namespace dump {

static void iter_messages(const dballe::Messages& msgs, std::function<void(const dballe::Msg&)> cb) {
    for (dballe::Messages::const_iterator i = msgs.begin(); i != msgs.end(); ++i)
        cb(dballe::Msg::downcast(*i));
}

static void iter_message(const dballe::Msg& msg, std::function<void(const dballe::msg::Context&)> cb) {
    for (std::vector<dballe::msg::Context*>::const_iterator i = msg.data.begin(); i != msg.data.end(); ++i)
        cb(**i);
}

static void iter_context(const dballe::msg::Context& ctx, std::function<void(const wreport::Var&)> cb) {
    for (std::vector<wreport::Var*>::const_iterator i = ctx.data.begin(); i != ctx.data.end(); ++i)
        cb(**i);
}

static void iter_var(const wreport::Var& var, std::function<void(const wreport::Var&)> cb) {
    for (const wreport::Var* attr = var.next_attr(); attr != NULL; attr = attr->next_attr())
        cb(*attr);
}

JsonDumper::JsonDumper(std::ostream& out) : out(out), json(false, NULL) {}

void JsonDumper::flush() {
    out << json.flush();
}

void GeoJsonDumper::start_dump() {
    json.start_map();
    json.add_string("type");
    json.add_string("FeatureCollection");
    json.add_string("features");
    json.start_list();
}

void GeoJsonDumper::end_dump() {
    json.end_list();
    json.end_map();
    flush();
}

bool GeoJsonDumper::operator()(const dballe::cmdline::Item& item) {
    iter_messages(*(item.msgs), [&](const dballe::Msg& msg) {
        iter_message(msg, [&](const dballe::msg::Context& ctx) {
            if (ctx.is_station())
                return;
            iter_context(ctx, [&](const wreport::Var& var) {
                json.start_map();
                json.add_string("type");
                json.add_string("Feature");
                json.add_string("geometry");
                json.start_map();
                json.add_string("type");
                json.add_string("Point");
                json.add_string("coordinates");
                json.start_list();
                if (msg.get_longitude_var())
                    json.add_number(msg.get_longitude_var()->format());
                else
                    json.add_null();
                if (msg.get_latitude_var())
                    json.add_number(msg.get_latitude_var()->format());
                else
                    json.add_null();
                json.end_list();
                json.end_map();
                json.add_string("properties");
                json.start_map();
                json.add_string("lon");
                json.add_int(msg.get_longitude_var()->enqi());
                json.add_string("lat");
                json.add_int(msg.get_latitude_var()->enqi());
                json.add_string("datetime");
                std::stringstream ss;
                msg.get_datetime().to_stream_iso8601(ss, 'T', "Z");
                json.add_string(ss.str().c_str());
                json.add_string("level_t1");
                json.add_int_or_null(ctx.level.ltype1);
                json.add_string("level_v1");
                json.add_int_or_null(ctx.level.l1);
                json.add_string("level_t2");
                json.add_int_or_null(ctx.level.ltype2);
                json.add_string("level_v2");
                json.add_int_or_null(ctx.level.l2);
                json.add_string("trange_pind");
                json.add_int_or_null(ctx.trange.pind);
                json.add_string("trange_p1");
                json.add_int_or_null(ctx.trange.p1);
                json.add_string("trange_p2");
                json.add_int_or_null(ctx.trange.p2);
                json.add_string("network");
                json.add_string(msg.get_rep_memo_var() ? msg.get_rep_memo_var()->enqc() : dballe::Msg::repmemo_from_type(msg.type));
                json.add_string("ident");
                if (msg.get_ident_var() != NULL)
                    json.add_string(msg.get_ident_var()->enqc());
                else
                    json.add_null();
                json.add_string("bcode");
                json.add_string(wreport::varcode_format(var.code()));
                json.add_string("value");
                json.add_var_value(var);
                json.end_map();
                json.end_map();
            });
        });
    });
    return true;
}

void DballeJsonDumper::start_dump() {
    json.start_list();
}

void DballeJsonDumper::end_dump() {
    json.end_list();
}

bool DballeJsonDumper::operator()(const dballe::cmdline::Item& item) {
    iter_messages(*(item.msgs), [&](const dballe::Msg& msg) {
        json.start_map();
        json.add_string("network");
        json.add_string(msg.get_rep_memo_var() ? msg.get_rep_memo_var()->enqc() : dballe::Msg::repmemo_from_type(msg.type));
        json.add_string("ident");
        if (msg.get_ident_var() != NULL)
            json.add_string(msg.get_ident_var()->enqc());
        else
            json.add_null();
        json.add_string("lon");
        json.add_int(msg.get_longitude_var()->enqi());
        json.add_string("lat");
        json.add_int(msg.get_latitude_var()->enqi());
        json.add_string("date");
        std::stringstream ss;
        msg.get_datetime().to_stream_iso8601(ss, 'T', "Z");
        json.add_string(ss.str().c_str());
        json.add_string("data");
        json.start_list();
        iter_message(msg, [&](const dballe::msg::Context& ctx) {
            json.start_map();
            if (not ctx.is_station()) {
                json.add_string("timerange");
                json.start_list();
                json.add_int_or_null(ctx.trange.pind);
                json.add_int_or_null(ctx.trange.p1);
                json.add_int_or_null(ctx.trange.p2);
                json.end_list();
                json.add_string("level");
                json.start_list();
                json.add_int_or_null(ctx.level.ltype1);
                json.add_int_or_null(ctx.level.l1);
                json.add_int_or_null(ctx.level.ltype2);
                json.add_int_or_null(ctx.level.l2);
                json.end_list();
            }
            json.add_string("vars");
            json.start_map();
            iter_context(ctx, [&](const wreport::Var& var) {
                json.add_string(wreport::varcode_format(var.code()));
                json.start_map();
                json.add_string("v");
                json.add_var_value(var);
                json.add_string("a");
                json.start_map();
                iter_var(var, [&](const wreport::Var& attr) {
                    json.add_string(wreport::varcode_format(attr.code()));
                    json.add_var_value(attr);
                });
                json.end_map();
                json.end_map();
            });
            json.end_map();
            json.end_map();
        });
        json.end_list();
        json.end_map();
    });
    return true;
}

}
}
